dcplib/etl/__init__.py

```diff
@@ -0,0 +1,128 @@
+                try:
+                    with open(f"{self.sd}/files/{f['uuid']}", "rb") as fh:
+                        file_csum = hashlib.sha256(fh.read()).hexdigest()
+                        if file_csum == f["sha256"]:
+                            logger.debug("Linking fetched file %s/%s", bundle_uuid, f["name"])
+                            if not os.path.exists(f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}"):
+                                os.symlink(f"../../files/{f['uuid']}",
+                                           f"{self.sd}/bundles/{bundle_uuid}.{bundle_version}/{f['name']}")
+                            continue
+                except (FileNotFoundError, ):
+                    pass
+                files_to_fetch.append(f)
+            else:
+                logger.debug("Skipping non-metadata file %s/%s", bundle_uuid, f["name"])
+        return bundle_uuid, bundle_version, files_to_fetch
+
+    def enqueue_bundle_manifests(self, percent_complete, bundles_complete, total_bundles, f2f_futures, ff_futures,
+                                 executor):
+        logger.info("[%d%%] [%d/%d] Processing bundle batch",
+                    percent_complete, bundles_complete, total_bundles)
+        for future in concurrent.futures.as_completed(f2f_futures):
+            bundle_uuid, bundle_version, files_to_fetch = future.result()
+            if files_to_fetch:
+                logger.info("[%d%%] Fetching %d files for bundle %s @ %s",
+                            percent_complete, len(files_to_fetch), bundle_uuid, bundle_version)
+                for f in files_to_fetch:
+                    ff_futures.append(executor.submit(self.get_file, f, bundle_uuid, bundle_version))
+                    self.b2f[bundle_uuid + "." + bundle_version].add(f["name"])
+            else:
+                self.staged_bundles.append([bundle_uuid, bundle_version])
+        return len(f2f_futures)
+
+    def extract(self, query: dict = None, transformer: callable = None, loader: callable = None,
+                content_type_patterns=None, filename_patterns=None, max_workers=512):
+        if query is None:
+            query = self.default_bundle_query
+        if content_type_patterns is None:
+            content_type_patterns = self.default_content_type_patterns
+        os.makedirs("files", exist_ok=True)
+        os.makedirs("bundles", exist_ok=True)
+        f2f_futures, ff_futures = [], []
+        bundles_complete, percent_complete = 0, -1
+        with concurrent.futures.ThreadPoolExecutor(max_workers=max_workers) as executor:
+            dss = hca.dss.DSSClient(swagger_url=f"{self.dss_host}/swagger.json")
+            total_bundles = dss.post_search(es_query=query, replica="aws")["total_hits"]
+            logging.info("Scanning %s bundles", total_bundles)
+            for b in dss.post_search.iterate(es_query=query, replica="aws", per_page=500):
```

Comment:
Does `post_search` implement `iterate`? Does this handle pagination or 500 results max?

Reply:
Yes. See the pagination support in dcp-cli. The 500 max set here is to maximize the number of results fetched per request, since we want to minimize the latency overhead.